docs: add AGENTS.md universal agent guide (#26018)
strapi/strapi: AGENTS.md

# Strapi Monorepo — Agent Guide

Strapi is an open-source headless CMS.
Yarn workspaces + Nx monorepo. Node ≥20 ≤24, Yarn 4.
Target branch: `develop` (not `main`). All PRs go to `develop`.

> Per-package `AGENTS.md` files are planned for future iterations — none exist yet.

Reviewer: What about removing this until it actually exists. Rationale: It's a non information for the agent reading it.

packages/cli/        # CLI tools: create-strapi-app, cloud-cli
examples/            # Dev sandboxes only — not published, not for production fixes
docs/                # Contributor documentation (also at contributor.strapi.io)
tests/               # Test infrastructure scripts
```

| Package                            | Description                                               |
| ---------------------------------- | --------------------------------------------------------- |
| `@strapi/strapi`                   | Main framework entry point (Koa server)                   |
| `@strapi/admin`                    | React 18 admin dashboard                                  |
| `@strapi/core`                     | Core business logic                                       |
| `@strapi/database`                 | Database abstraction (MySQL, PostgreSQL, MariaDB, SQLite) |
| `@strapi/content-manager`          | Content management UI                                     |
| `@strapi/types`                    | Shared TypeScript type definitions                        |
| `@strapi/permissions`              | RBAC engine                                               |

> | `@strapi/plugin-users-permissions` | JWT authentication                                        |

Reviewer: Since we have 37 scoped packages and many unscoped. What about getting rid of this list and rather rely on package README for description?
OR
We indicate that these are the most important packages but the list is not exhaustive + rationale + way to get more details.

Scoped:
```
@strapi/admin, @strapi/admin-test-utils, @strapi/cloud-cli, @strapi/content-manager, @strapi/content-releases, @strapi/content-type-builder, @strapi/core, @strapi/data-transfer, @strapi/database, @strapi/e2e-test-app, @strapi/email, @strapi/generators, @strapi/i18n, @strapi/logger, @strapi/openapi, @strapi/permissions, @strapi/plugin-cloud, @strapi/plugin-color-picker, @strapi/plugin-documentation, @strapi/plugin-graphql, @strapi/plugin-sentry, @strapi/plugin-users-permissions, @strapi/provider-email-amazon-ses, @strapi/provider-email-mailgun, @strapi/provider-email-nodemailer, @strapi/provider-email-sendgrid, @strapi/provider-email-sendmail, @strapi/provider-upload-aws-s3, @strapi/provider-upload-cloudinary, @strapi/provider-upload-local, @strapi/review-workflows, @strapi/strapi, @strapi/types, @strapi/typescript-utils, @strapi/upgrade, @strapi/upload, @strapi/utils
```

Unscoped:
```
api, api-tests, check-pr-status, complex, create-strapi, create-strapi-app, docs, empty, eslint-config-custom, experimental-dev, getstarted, kitchensink, kitchensink-ts, scripts-front, strapi-plugin-local-plugin, strapi-plugin-todo-example, strapi-plugin-workspace-plugin, tsconfig, vitest-config,
```

---

> ## Architecture

Reviewer 2: This list might be lacking a bit of direction. Some things dont quite feel like architecture relevant topics.

> - **`strapi` global** — The `Strapi` class is the DI container and central hub, accessible as `strapi` everywhere (e.g. `strapi.documents`, `strapi.db`, `strapi.log`). Lifecycle: Register → Bootstrap → Start → Destroy (start() calls load() internally, which runs register + bootstrap).

Reviewer: I'm afraid that "`strapi` global" could be interpreted as `global.strapi` that we must avoid to use at all costs. We always prefer proper DI through factory pattern: `createService(strapi)`.

> - **Server / Admin split** — Koa.js HTTP server (`@strapi/strapi`) + React/Redux admin (`@strapi/admin`). Packages with both concerns export dual entry points: `strapi-server` (Node.js logic) and `strapi-admin` (UI components). Changes to one rarely require changes to the other.

Reviewer: > Changes to one rarely require changes to the other.

I feel like this is not really accurate. It can be true for fixes but rarely true for new features or large refactor. I'd just get rid of that mention.

- **Document Service** — The primary high-level API for content (`strapi.documents`). Replaced the legacy Entity Service. Always use this for reading/writing content — never raw DB queries unless you're working inside `@strapi/database` itself.

> - **Plugin system** — Plugins register routes, controllers, services, content types, and middleware via the same `strapi-server` / `strapi-admin` dual structure. Official plugins live in `packages/plugins/`; app-level plugins in `src/plugins/`.

Reviewer: > app-level plugins in `src/plugins/`

Here we're referring to `app-level` a concept we do not explain elsewhere. 
I'd either delete the mention or create a glossary section to expand on those terms.

> - **Content Types** — Defined as JSON schemas. The database layer auto-generates tables from them. Never write raw migrations for content type changes.

Reviewer: > Defined as JSON schemas

I'm a bit afraid this gets confused with actual JSON schema spec which is foreign of our definition. Maybe "Defined using a JSON based notation".

> - **Content Types** — Defined as JSON schemas. The database layer auto-generates tables from them. Never write raw migrations for content type changes.

Reviewer 2: Should we add an example of what it looks like or add a link to the documentation? 
We could also add an indication of where it lives in the folder structure.

> - **Draft & Publish** — Core content workflow. Documents can exist in draft or published state; the Document Service handles the state machine.

Reviewer: IMO this goes into a glossary or features section

- **EE / CE split** — Some features are Enterprise Edition only, gated at runtime. See EE toggles in Notes for Agents.
- **`@strapi/types`** — Single source of truth for shared TypeScript types. Import from here; improve these types rather than duplicating locally.

---

> ## Setup & Development

Reviewer: ```suggestion
## Monorepo setup
```

```bash
# Initial setup (run once after cloning)
yarn install
yarn setup                    # clean + build all packages

# Start databases (postgres/mysql)
docker-compose -f docker-compose.dev.yml up -d

# Run the dev sandbox
cd examples/getstarted
yarn develop                  # SQLite (default)
DB=postgres yarn develop      # PostgreSQL
DB=mysql yarn develop         # MySQL

# Watch all packages + run sandbox with admin watch (two terminals)
yarn watch                                           # terminal 1, repo root

> cd examples/getstarted && yarn develop --watch-admin # terminal 2

Reviewer: What about splitting and creating a dedicated section

## Development

```bash

# Run the dev sandbox
cd examples/getstarted
yarn develop                  # SQLite (default)

# Start non-memory databases (postgres/mysql)
docker-compose -f docker-compose.dev.yml up -d
# run
DB=postgres yarn develop      # PostgreSQL
DB=mysql yarn develop         # MySQL

# Watch all packages + run sandbox with admin watch (two terminals)
yarn watch                                           # terminal 1, repo root
cd examples/getstarted && yarn develop --watch-admin # terminal 2
```

> # Build

Reviewer: ## Build

```bash
…

yarn build              # all packages (code + types)
yarn build:code         # faster — skips .d.ts generation
yarn nx build @strapi/admin   # single package
```

---

> ## Testing

Reviewer: We are missing CLI tests

> ## Testing

Reviewer 2: is there an option tu run the whole suite? yarn test?

Reviewer: You mean running all tests at once @oiorain? If so, this is not something the repo is ready for. Only the E2E cannot be launched all at once without crashing the memory of its host, running into timeouts otherwise.

```bash
# Unit (fastest — run first)
yarn test:unit
yarn test:unit:watch

# Frontend (admin panel)
yarn test:front              # runs with IS_EE=true (EE features enabled)
yarn test:front:ce           # runs with IS_EE=false (Community Edition only)

> 

Reviewer 3: since you added the update snapshots for the api tests, you can maybe add it here for the unit tests:

```suggestion
yarn test:front --updateSnapshot          # update snapshots
```

# Type checking
yarn test:ts                 # all packages + front + back

# API integration (auto-generates a test app)
yarn test:api                # SQLite
yarn test:api --db=postgres
yarn test:api --db=mysql
yarn test:api -u             # update snapshots

# E2E (Playwright)
yarn playwright install      # one-time browser install

> yarn test:e2e --setup --concurrency=1

Reviewer 4: would be nice if we have more details here on how to run test:e2e with certain browsers and a specific domain and file

```

**Before every PR:**

```bash

> yarn test:unit && yarn test:front && yarn test:ts && yarn lint

Reviewer: ```suggestion
yarn test:unit && yarn test:front && yarn test:ts && yarn lint && yarn prettier:check
```

or even

```suggestion
yarn test:unit && yarn test:front && yarn test:ts && yarn lint:fix && yarn prettier:write
```

```

E2E (`yarn test:e2e`) is required per CONTRIBUTING.md but slow — CI enforces it on every PR.

> **Test file location:** `__tests__/` subdirectories within each package (not a root `tests/` folder).

Reviewer: I feel like this should be distributed in "# Unit (fastest — run first)", "# Frontend (admin panel)", "# API integration (auto-generates a test app)", "# E2E (Playwright)".
A one liner for each surfacing the diff between close-to-code BE unit / FE unit in `__tests__` and the root `./tests/`

> **When to add tests:** Always for bug fixes (reproduce the bug first). For features, add tests when they cover meaningful behaviour — not just to hit coverage numbers.

Reviewer 3: Might be worth adding that sometimes it's not just about adding tests but updating them?

---

## Quality Gates

> ### Commits

Reviewer 3: Might be worth adding a sentence specifying that the commit should pass all tests before merging.

> Must pass `commitlint` (enforced by Husky and CI). Format:

Reviewer: `commitlint` is not part of package.json scripts (only `commit` is). Not sure how we should instruct how to run it outside of husky.

> Must pass `commitlint` (enforced by Husky and CI). Format:

Reviewer: `commitlint` check is actually against: https://www.conventionalcommits.org/en/v1.0.0/ maybe worth mentioning

```
<type>(<optional-scope>): <description>
```

Valid types: `feat` `fix` `chore` `ci` `docs` `enhancement` `test` `revert` `security` `future`

> 

Reviewer: ```suggestion

**Examples:**

```

```bash
feat(content-manager): add bulk delete action
fix(database): preserve relation order during publish
chore(admin): migrate data-fetching to react-query
```

Use `yarn commit` for an interactive prompt. Run `yarn version:check` if you've touched any `package.json`.

### CLA

> First-time contributors must sign the [Contributor License Agreement](https://cla.strapi.io/strapi/strapi) before a PR can be merged. The CLA bot prompts automatically on the PR.

Reviewer: Do we actually want this to instruct LLMs since only humans can do the action?
Rationale: it's a README / CONTRIBUTING scope

### TypeScript

- Import types from `@strapi/types` — extend or improve them rather than duplicating locally.

> - Never use `any` when a proper type exists or can be reasonably defined.

Reviewer: ```suggestion
- Never use `any` when a proper type exists or can be reasonably defined. Prefer `unknown` otherwise.
```

> - Run `yarn test:ts` before pushing.

Reviewer: Don't we prefer a full build here? No opinion.

### Linting & Formatting

```bash
yarn lint           # ESLint across all packages
yarn lint:fix       # auto-fix

> yarn format         # Prettier (2-space indent, single quotes, semicolons, trailing commas, arrow parens, 100-char width, LF)

Reviewer: oh! I never noticed we have both `prettier:*` and `format`! 
I personally prefer format but not sure we can get rid of the rest.

yarn prettier:check # check only
```

---

## PR Guidelines

- Branch from `develop`, target `develop` — never `main`.
- Link the issue you're fixing in the description.
- PR description must use this template:
  - **What does it do?** — technical changes made
  - **Why is it needed?** — problem being solved
  - **How to test it?** — steps to reproduce and verify
  - **Related issue(s)/PR(s)** — links

---

## Notes for Agents

> - **`examples/`** apps are sandboxes only — use them to reproduce and test fixes, never commit production changes to them.

Reviewer: ```suggestion
- **`examples/`** apps are sandboxes only — use them to reproduce and test fixes, never commit changes to them, unless specifically asked to do so.
```

> - **Test files** belong in `__tests__/` subdirectories, not a root `tests/` folder.

Reviewer: true for unit, false for integration / E2E. 
I'd rather remove the mention from here.

- **Workspace deps** — internal `packages/` deps reference each other with pinned semver versions (e.g. `"5.42.0"`), not `workspace:*`. The `workspace:*` protocol is only used in `examples/` apps and some root devDeps.
- **Entity Service is deprecated** — always use the Document Service (`strapi.documents`) for content operations.
- **Lifecycle phases** — `strapi.isLoaded` must be `true` before accessing services. Plugins and DB are not available until after the `load()` phase.

> - **API test apps are ephemeral** — regenerate with `yarn test:generate-app` rather than reusing a stale one; stale apps cause misleading test failures.

Reviewer: I feel this needs more details. My guess: "test apps" refers to the test apps generated by cli / integration / E2E setup. Better to create a section in testing above.

> - **EE toggles** — `IS_EE=true` for frontend tests, `RUN_EE=true` for E2E tests.

Reviewer: Same for this: goes in testing section

# Strapi Monorepo — Agent Guide

> Strapi is an open-source headless CMS.

Reviewer 5: ```suggestion
This is the monorepo for the open-source headless CMS Strapi
```

- **Plugin system** — Plugins register routes, controllers, services, content types, and middleware via the same `strapi-server` / `strapi-admin` dual structure. Official plugins live in `packages/plugins/`.
- **Content Types** — Defined using a JSON-based notation (not JSON Schema spec). Each content type has a `schema.json` file — see any `packages/core/content-manager/server/src/content-types/` for examples. The database layer auto-generates tables from them. Never write raw migrations for content type changes.
- **EE / CE split** — Some features are Enterprise Edition only, gated at runtime. See EE toggles in the Testing section below.
- **`@strapi/types`** — Single source of truth for shared TypeScript types. Import from here; improve these types rather than duplicating locally.

---

## Monorepo Setup

```bash
# Initial setup (run once after cloning)
yarn install
yarn setup                    # clean + build all packages
```

---

## Development

```bash
# Run the dev sandbox

> cd examples/getstarted

Reviewer 5: should we replace this with complex since it's typescript?

Reviewer: Or we convert getstarted in TS?
Do you recommend we primarily use the complex project for local dev @innerdvations?

Reviewer 5: personally I would aim to eventually delete getstarted (even though I'm quite fond of it, it's been there my entire tenure at Strapi 😂) and just have 'complex' and 'empty'

```bash
yarn test:unit
yarn test:unit:watch
yarn test:unit:update         # update snapshots
```

### Frontend tests (admin panel)

Frontend test files also live in `__tests__/` within their respective packages.

```bash
yarn test:front              # runs with IS_EE=true (EE features enabled)
yarn test:front:ce           # runs with IS_EE=false (Community Edition only)
yarn test:front:update       # update snapshots (EE)
yarn test:front:update:ce    # update snapshots (CE)
```

### Type checking

```bash
yarn test:ts                 # all packages + front + back
```

### API integration tests

Integration tests live in `tests/api/`. Test apps are generated automatically — always regenerate with `yarn test:generate-app` rather than reusing a stale one (stale apps cause misleading failures).

```bash
yarn test:api                # SQLite
yarn test:api --db=postgres
yarn test:api --db=mysql
yarn test:api -u             # update snapshots
```

### CLI tests

CLI tests live in `tests/cli/`.

```bash
yarn test:cli
yarn test:cli:debug          # with debug output
yarn test:cli:update         # update snapshots
```

> ### E2E tests (Playwright)

Reviewer 5: Depending on which gets merged first, this will need to be updated with changes from https://github.com/strapi/strapi/pull/26019

---

## Quality Gates

### Commits

Must follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) (enforced by Husky `commit-msg` hook and CI via `commitlint`). Format:

```
<type>(<optional-scope>): <description>
```

Valid types: `feat` `fix` `chore` `ci` `docs` `enhancement` `test` `revert` `security` `future` `release`

**Examples:**

```bash
feat(content-manager): add bulk delete action
fix(database): preserve relation order during publish
chore(admin): migrate data-fetching to react-query
```

Use `yarn commit` for an interactive prompt. Run `yarn version:check` if you've touched any `package.json`.

### TypeScript

- Import types from `@strapi/types` — extend or improve them rather than duplicating locally.
- Never use `any` when a proper type exists or can be reasonably defined. Prefer `unknown` otherwise.
- Run `yarn test:ts` before pushing.

### Linting & Formatting

```bash
yarn lint           # ESLint across all packages
yarn lint:fix       # auto-fix
yarn format         # Prettier (2-space indent, single quotes, semicolons, trailing commas, arrow parens, 100-char width, LF)
yarn prettier:check # check only
```

---

> ## Security

Reviewer 5: Just curious, are many people giving AI full control over automatically creating commits and PRs that this is useful?

- Never commit secrets, credentials, or API keys.
- Never disable or weaken authentication/authorization checks.
- Use parameterized queries — never interpolate user input into raw SQL or database queries.
- Validate and sanitize all user input at controller/service boundaries.
- When working with EE-gated features, do not bypass license checks.

---

## PR Guidelines

> - Branch from `develop`, target `develop` — never `main`.

Reviewer 5: Duplicates the line at the top

- Link the issue you're fixing in the description.
- All tests must pass before merging.
- PR description must use this template:
  - **What does it do?** — technical changes made
  - **Why is it needed?** — problem being solved
  - **How to test it?** — steps to reproduce and verify
  - **Related issue(s)/PR(s)** — links

---

## Notes for Agents

- **`examples/`** apps are sandboxes only — use them to reproduce and test fixes, never commit changes to them unless specifically asked to do so.

> - **Workspace deps** — internal `packages/` deps reference each other with pinned semver versions (e.g. `"5.42.0"`), not `workspace:*`. The `workspace:*` protocol is only used in `examples/` apps and some root devDeps.

Reviewer 6: @butcherZ I always wondered, Why Strapi `packages/` are not using `workspace:*` protocol? Are there any reason why it needs to be a fixed version?

From what I recall, on release pnpm/yarn automatically changes the `workspace:*` protocol to the actual release version.

cc @innerdvations @nclsndr

Reviewer 6: PS: Will a file for Claude Code be added because is "special" 👀 

Related: https://github.com/strapi/strapi/pull/26251
